Feat oauth refresh tokens

Meldiron opened the pull request:
## What does this PR do?

🚨 BREAKING CHANGES 🚨
`providerToken` in database is now `providerAccessToken`. Same in response model
🚨🚨🚨🚨🚨🚨

This PR stores the expiry token and expiryDate on a session if OAuth authentication is used. This is really powerful for developers since access tokens are short-lived and would require users to re-login every 4 hours.

TODO: This also creates `PATCH /v1/account/sessions/{sessionId}/oauth-tokens` endpoint to refresh token after expiration.

## Test Plan

I added a test for Mock OAuth2 covering new`updateSession` endpoint.

## Related PRs and Issues


### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?

✅

eldadfux commented on app/config/collections.php:
```diff
@@ -1158,7 +1158,18 @@
+                'filters' => ['encrypt'],
+            ],
+            [
+                '$id' => 'providerRefreshToken',
```
Let's add `providerAccessTokenExpiry` with a unix timestamp as the value.

Meldiron replied:
Done.

eldadfux commented on src/Appwrite/Auth/OAuth2.php:
```diff
@@ -62,9 +62,9 @@ abstract public function getLoginURL():string;
+     * @return array
      */
-    abstract public function getAccessToken(string $code):string;
+    abstract public function getTokens(string $code):array;
```
A better developer friendly signature might be to have both `getAccessToken` and `getRefreshToken` methods where each returns a string. We can store the result of the first API call and cache it in `private $tokens` if `$this->tokens` is not empty we can skip a second API call and just return the relevant string according to the method needs.

We follow a similar pattern for `getUserEmail`, `getUserID`, and `getUserName`.

Meldiron replied:
User's actually have `getUser` that gets the data, and functions `getUserEmail` `getUserName`to return specific ones.

To follow that, we would have `getTokens`, store the response in variable, and then return specific parts with functions `getAccessToken` and `getRefreshToken`.

Hope that is what you meant. I have implemented in Sporify, please let me know what you think, and I can add it elsewhere: https://github.com/appwrite/appwrite/pull/2733/files#diff-00ac4df20657d38a00b1d590988c2daee3ba5476f65b9b915ef3851cbc453deb

(getRefreshToken, getTokenExpiry and getAccessToken will be in abstract class to prevent repetition)

eldadfux replied:
Yes, this example look good, but `getTokens` can be a private method.

Meldiron replied:
Changed it to 'protected'. I added some universal functions to OAuth class, and it now needs to access this `getTokens` method.

eldadfux reviewed (commented):
TODO:

- [ ] Add access token expiry
- [ ] Add endpoint for refreshing the endpoint, what should it look like? Maybe PATCH request for the given session?

brandonroberts commented on src/Appwrite/Utopia/Response/Model/Session.php:
```diff
@@ -40,9 +40,21 @@ public function __construct()
+            ])
+            ->addRule('providerAccessTokenExpiry', [
+                'type' => self::TYPE_INTEGER,
+                'description' => 'Date, the user has accepted the invitation to join the team in Unix timestamp.',
```
```suggestion
                'description' => 'Date, the Unix timestamp of when the access token expires',
```

brandonroberts commented on src/Appwrite/Auth/OAuth2/Amazon.php:
```diff
@@ -59,31 +64,27 @@ public function getLoginURL(): string
-            return $accessToken['access_token'];
+        if(empty($this->tokens)) {
+            $headers = ['Content-Type: application/x-www-form-urlencoded;charset=UTF-8'];
+            $this->tokens = \json_decode($this->request(
```
Just so I understand correctly, this stores the whole response from the access token request right? But we only provide ways to get the access token and refresh token publicly?

How does this handle nested JSON objects?

Meldiron replied:
Yes, methods `getRefreshToken` and `getAccessToken` are on parent - on OAuth2.php. I decided to put it there because all of our current providers follow the same response structure. If there is some provider that does not, the function could be overwritten.

brandonroberts commented on src/Appwrite/Auth/OAuth2/Notion.php:
```diff
@@ -51,32 +56,25 @@ public function getLoginURL():string
-        if (isset($response['access_token'])) {
-            return $response['access_token'];
+        if(empty($this->tokens)) {
+            $headers = ['Authorization: Basic ' . \base64_encode($this->appID . ':' . $this->appSecret)];
```
Should building the `Authorization Basic` header become a function for reuse?

Meldiron replied:
It could, for this PR I re-used what was already there. We could make new PR with these improvements, I also saw a lo of potential for re-use. 90% of providers follow protocol strictly and only differs in path.

Meldiron commented:
- Added tokenExpiry information

![CleanShot 2022-02-01 at 16 34 28](https://user-images.githubusercontent.com/19310830/151999426-b56f45e1-6659-4565-8408-3070f745139c.png)

- Implemented `sdk.account.updateOAuth2Tokens("current", true)`

Code:

```
   async function flow() {
    var { providerAccessTokenExpiry: e, providerAccessToken: t } =
     await sdk.account.getSession("current");
    console.log("Current token ", t, " , expires at ", e);

    var { providerAccessTokenExpiry: e, providerAccessToken: t } =
     await sdk.account.updateOAuth2Tokens("current");
    console.log("Updated token to ", t, " , expires at ", e);

    var { providerAccessTokenExpiry: e, providerAccessToken: t } =
     await sdk.account.getSession("current");
    console.log("Current token ", t, " , expires at ", e);

    var { providerAccessTokenExpiry: e, providerAccessToken: t } =
     await sdk.account.updateOAuth2Tokens("current", true);
    console.log("FORCE Updated token to ", t, " , expires at ", e);

    var { providerAccessTokenExpiry: e, providerAccessToken: t } =
     await sdk.account.getSession("current");
    console.log("Current token ", t, " , expires at ", e);
   }
```

Output:
![CleanShot 2022-02-01 at 17 36 29](https://user-images.githubusercontent.com/19310830/152010951-f34ec411-f1e4-465b-867c-0da9a80ce20b.png)

(first-time token was not re-generated and responded with empty 204, because it did not expire yet. Second time I forced it with `true`, so new token was provided)

Meldiron commented:
> TODO:
> 
> * [ ]  Add access token expiry
> * [ ]  Add endpoint for refreshing the endpoint, what should it look like? Maybe PATCH request for the given session?

Both addressed

eldadfux commented on app/controllers/api/account.php:
```diff
@@ -1647,6 +1654,99 @@
+
+        $sessions = $user->getAttribute('sessions', []);
+
+        foreach ($sessions as $key => $session) {/** @var Document $session */
```
I can definitely see your problem with this. We could try and optimize the code, but could also consider deprecating the sessions attribute and just use a subquery like we do for the project sub-entities.

Meldiron replied:
I checked the rest of the code and I believe we currently do a lot of duplicate work. Every write is done on both 'sessions' and 'user'. We 100% should have subQuery on sessions, but that would need to be a separate refactor.

With that said, I would stick with this solution that can get slower with many active sessions, and fix it alongside other endpoints once we introduce subQuery for sessions.

eldadfux replied:
Agree, and it will be better to address it in a new PR.

eldadfux commented on app/controllers/api/account.php:
```diff
@@ -1647,6 +1654,99 @@
         throw new Exception('Session not found', 404);
     });
 
+App::patch('/v1/account/sessions/:sessionId/oauth2-tokens')
```
I'm not sure this path makes total sense. I'd love us to come with other REST complaint ideas.

Meldiron replied:
Changed to `PATCH /v1/account/sessions/:sessionId`, and we can use same endpoint for session renewing in the future.

eldadfux commented on src/Appwrite/Auth/OAuth2.php:
```diff
@@ -62,9 +62,16 @@ abstract public function getLoginURL():string;
+     * @return array
      */
-    abstract public function getAccessToken(string $code):string;
+    abstract public function refreshTokens(string $refreshToken):array;
```
`refreshTokens` or `refreshToken`?

Meldiron replied:
Tokens. Auth0 says `With refresh token rotation enabled, every time an application exchanges a refresh token to get a new access token, a new refresh token is also returned `.

This means refresh token can also change when the access token is renewed, so one or two tokens could be changing, depending on specific OAuth provider.

eldadfux commented on src/Appwrite/Auth/OAuth2.php:
```diff
@@ -109,6 +116,38 @@ protected function getScopes():array
+    public function getAccessToken(string $code):string
+    {
+        $tokens = $this->getTokens($code);
+        return $tokens['access_token'];
```
We could use null safety here to make the code a bit more solid in case the response hasn't came in as expected.

Meldiron replied:
Adde null safety

eldadfux commented on src/Appwrite/Auth/OAuth2.php:
```diff
@@ -109,6 +116,38 @@ protected function getScopes():array
+    public function getRefreshToken(string $code):string
+    {
+        $tokens = $this->getTokens($code);
+        return $tokens['refresh_token'];
```
We could use null safety here to make the code a bit more solid in case the response hasn't came in as expected.

Meldiron replied:
Added null safety

eldadfux commented on src/Appwrite/Auth/OAuth2.php:
```diff
@@ -109,6 +116,38 @@ protected function getScopes():array
+    public function getAccessTokenExpiry(string $code):string
+    {
+        $tokens = $this->getTokens($code);
+        return $tokens['expires_in'];
```
We could use null safety here to make the code a bit more solid in case the response hasn't came in as expected.

Meldiron replied:
Adde null safety

eldadfux commented on src/Appwrite/Utopia/Response/Model/Session.php:
```diff
@@ -40,9 +40,21 @@ public function __construct()
+            ])
+            ->addRule('providerAccessTokenExpiry', [
+                'type' => self::TYPE_INTEGER,
+                'description' => 'Date, the Unix timestamp of when the access token expires',
```
```suggestion
                'description' => 'Date, the Unix timestamp of when the access token expires.',
```

eldadfux commented on app/controllers/api/account.php:
```diff
@@ -1647,6 +1654,99 @@
+            if ($sessionId == $session->getId()) {
+                $expireAt = (int) $session->getAttribute('providerAccessTokenExpiry');
+                if(\time() < $expireAt - 5 && !$force) { // 5 seconds time-sync and networking gap, to be safe
+                    return $response->noContent();
```
Not sure this is the right response here. We usually use no-content only on delete operations.

Meldiron replied:
Added null safety

Meldiron replied:
I updated this part by removing the force parameter. If a developer decides to hit this endpoint, we always renew the token. They can check expiryDate from session to prevent firing it too often. Worst-case scenario, their app ID/secret hits the API limit. Not a concern for Appwrite Cloud.

eldadfux commented on app/controllers/api/account.php:
```diff
@@ -1647,6 +1654,99 @@
+                $appSecret = $project->getAttribute('providers', [])[$provider.'Secret'] ?? '{}';
+
+                $className = 'Appwrite\\Auth\\OAuth2\\'.\ucfirst($provider);
+                $oauth2 = new $className($appId, $appSecret, '', [], []);
```
We're missing a check to see if provider is actually supported.

```php
        $className = 'Appwrite\\Auth\\OAuth2\\' . \ucfirst($provider);
        if (!\class_exists($className)) {
            throw new Exception('Provider is not supported', 501);
        }
```

Meldiron replied:
Great catch, thanks.

eldadfux commented on app/controllers/api/account.php:
```diff
@@ -1647,6 +1654,103 @@
     });
 
+App::patch('/v1/account/sessions/:sessionId')
+    ->desc('Update session - refresh OAuth2 access token')
```
```suggestion
    ->desc('Update Session (Refresh Tokens)')
```
